File: src/main/java/komu/blunt/analyzer/AnalyzingVisitor.java
package komu.blunt.analyzer;

import komu.blunt.ast.*;
import komu.blunt.core.*;
import komu.blunt.objects.Symbol;
import komu.blunt.objects.TypeConstructorValue;
import komu.blunt.types.ConstructorDefinition;
import komu.blunt.types.DataTypeDefinitions;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;
import static komu.blunt.objects.Symbol.symbol;

final class AnalyzingVisitor implements ASTVisitor<StaticEnvironment, CoreExpression> {

    private final DataTypeDefinitions dataTypes;
    private final PatternAnalyzer patternAnalyzer = new PatternAnalyzer();
    private int sequence = 1;

    public AnalyzingVisitor(DataTypeDefinitions dataTypes) {
        this.dataTypes = checkNotNull(dataTypes);
    }

    public CoreExpression analyze(ASTExpression exp, StaticEnvironment env) {
        return exp.accept(this, env);
    }

    private List<CoreExpression> analyzeAll(List<ASTExpression> exps, StaticEnvironment env) {
        List<CoreExpression> result = new ArrayList<>(exps.size());

        for (ASTExpression exp : exps)
            result.add(analyze(exp, env));

        return result;
    }

    @Override
    public CoreExpression visit(ASTApplication application, StaticEnvironment env) {
        return new CoreApplicationExpression(analyze(application.func, env), analyze(application.arg, env));
    }

    @Override
    public CoreExpression visit(ASTConstant constant, StaticEnvironment env) {
        return new CoreConstantExpression(constant.value);
    }

    @Override
    public CoreExpression visit(ASTLambda lambda, StaticEnvironment env) {
        StaticEnvironment newEnv = env.extend(lambda.argument);
        CoreExpression body = analyze(lambda.body, newEnv);
        return new CoreLambdaExpression(newEnv.size(), body);
    }

    @Override
    public CoreExpression visit(ASTLet let, StaticEnvironment env) {
        if (let.bindings.size() != 1)
            throw new UnsupportedOperationException("multi-var let is not supported");

        ImplicitBinding binding = let.bindings.get(0);
        VariableReference var = env.define(binding.name);

        CoreExpression expr = analyze(binding.expr, env);
        CoreExpression body = analyze(let.body, env);
        return new CoreLetExpression(var, expr, body);
    }

    @Override
    public CoreExpression visit(ASTLetRec let, StaticEnvironment env) {
        if (let.bindings.size() != 1)
            throw new UnsupportedOperationException("multi-var let is not supported");

        ImplicitBinding binding = let.bindings.get(0);
        VariableReference var = env.define(binding.name);

        CoreExpression expr = analyze(binding.expr, env);
        CoreExpression body = analyze(let.body, env);
        return new CoreLetExpression(var, expr, body);
    }

    @Override
    public CoreExpression visit(ASTSequence sequence, StaticEnvironment env) {
        if (sequence.exps.isEmpty()) throw new AnalyzationException("empty sequence");

        return new CoreSequenceExpression(analyzeAll(sequence.exps, env));
    }

    @Override
    public CoreExpression visit(ASTSet set, StaticEnvironment env) {
        return new CoreSetExpression(env.lookup(set.var), analyze(set.exp, env));
    }

    @Override
    public CoreExpression visit(ASTVariable variable, StaticEnvironment env) {
        return new CoreVariableExpression(env.lookup(variable.var));
    }

    @Override
    public CoreExpression visit(ASTConstructor constructor, StaticEnvironment ctx) {
        ConstructorDefinition ctor = dataTypes.findConstructor(constructor.name);

        if (ctor.arity == 0)
            return AST.constant(new TypeConstructorValue(ctor.index, ctor.name)).accept(this, ctx);
        else
            return AST.variable(ctor.name).accept(this, ctx);
    }

    @Override
    public CoreExpression visit(ASTCase astCase, StaticEnvironment env) {
        CoreExpression exp = analyze(astCase.exp, env);

        Symbol var = symbol("$match" + sequence++);
        VariableReference matchedObject = env.define(var);
        CoreExpression body = createAlts(matchedObject, astCase.alternatives, env);
        return new CoreLetExpression(matchedObject, exp, body);
    }
    
    private CoreExpression createAlts(VariableReference matchedObject, List<ASTAlternative> alts, StaticEnvironment env) {
        if (alts.isEmpty())
            return analyze(AST.apply(AST.variable("error"), AST.constant("match failure")), env);
            
        ASTAlternative head = alts.get(0);
        List<ASTAlternative> tail = alts.subList(1, alts.size());

        CoreAlternative alt = analyze(head, matchedObject, env);
        return new CoreIfExpression(alt.extractor, alt.body, createAlts(matchedObject, tail, env));
    }

    private CoreAlternative analyze(ASTAlternative alt, VariableReference matchedObject, StaticEnvironment env) {
        CoreExpression extractor = patternAnalyzer.createExtractor(alt.pattern, matchedObject, env);
        CoreExpression body = analyze(alt.value, env);
        return new CoreAlternative(extractor, body);
    }
}

File: src/main/kotlin/komu/blunt/analyzer/AnalyzingVisitor.kt
package komu.blunt.analyzer

import komu.blunt.ast.*
import komu.blunt.core.*
import komu.blunt.objects.Symbol
import komu.blunt.objects.TypeConstructorValue
import komu.blunt.types.DataTypeDefinitions

class AnalyzingVisitor(val dataTypes: DataTypeDefinitions) {

    private val patternAnalyzer = PatternAnalyzer()
    private var sequence = 1

    fun analyze(exp: ASTExpression, env: StaticEnvironment): CoreExpression =
      when (exp) {
          is ASTConstant    -> CoreConstantExpression(exp.value)
          is ASTSet         -> CoreSetExpression(env.lookup(exp.variable), analyze(exp.exp, env))
          is ASTApplication -> visit(exp, env)
          is ASTLambda      -> visit(exp, env)
          is ASTLet         -> visit(exp, env)
          is ASTLetRec      -> visit(exp, env)
          is ASTSequence    -> visit(exp, env)
          is ASTVariable    -> visit(exp, env)
          is ASTConstructor -> visit(exp, env)
          is ASTCase        -> visit(exp, env)
          else              -> throw AnalyzationException("unknown expression '$exp'")
      }

    private fun visit(lambda: ASTLambda, env: StaticEnvironment): CoreExpression {
        val newEnv = env.extend(lambda.argument)
        val body = analyze(lambda.body, newEnv)
        return CoreLambdaExpression(newEnv.size, body)
    }

    private fun visit(application: ASTApplication, env: StaticEnvironment): CoreExpression =
        CoreApplicationExpression(analyze(application.func, env), analyze(application.arg, env))

    private fun visit(constructor: ASTConstructor, ctx: StaticEnvironment): CoreExpression {
        val ctor = dataTypes.findConstructor(constructor.name)

        return if (ctor.arity == 0)
            analyze(AST.constant(TypeConstructorValue(ctor.index, ctor.name)), ctx)
        else
            analyze(AST.variable(ctor.name), ctx)
    }

    private fun visit(sequence: ASTSequence, env: StaticEnvironment): CoreExpression =
        if (!sequence.exps.empty)
            CoreSequenceExpression(sequence.exps.map { analyze(it, env) })
        else
            throw AnalyzationException("empty sequence")

    private fun visit(let: ASTLet, env: StaticEnvironment): CoreExpression {
        if (let.bindings.size != 1)
            throw UnsupportedOperationException("multi-var let is not supported")

        val binding = let.bindings.first()
        val v = env.define(binding.name)

        val expr = analyze(binding.expr, env)
        val body = analyze(let.body, env)

        return CoreLetExpression(v, expr, body)
    }

    private fun visit(let: ASTLetRec, env: StaticEnvironment): CoreExpression {
        if (let.bindings.size != 1)
            throw UnsupportedOperationException("multi-var letrec is not supported")

        val binding = let.bindings.first()
        val v = env.define(binding.name)

        val expr = analyze(binding.expr, env)
        val body = analyze(let.body, env)

        return CoreLetExpression(v, expr, body)
    }

    private fun visit(variable: ASTVariable, env: StaticEnvironment): CoreExpression =
        CoreVariableExpression(env.lookup(variable.name))

    private fun visit(astCase: ASTCase, env: StaticEnvironment): CoreExpression {
        val exp = analyze(astCase.exp, env)

        val v = Symbol("\$match${sequence++}")
        val matchedObject = env.define(v)
        val body = createAlts(matchedObject, astCase.alternatives, env)
        return CoreLetExpression(matchedObject, exp, body)
    }

    private fun createAlts(matchedObject: VariableReference, alts: List<ASTAlternative>, env: StaticEnvironment): CoreExpression =
        if (alts.empty)
            analyze(AST.error("match failure"), env)
        else {
            val (extractor, body) = analyzeAlternative(alts.first(), matchedObject, env)
            CoreIfExpression(extractor, body, createAlts(matchedObject, alts.tail, env))
        }

    private fun analyzeAlternative(alt: ASTAlternative, matchedObject: VariableReference, env: StaticEnvironment): CoreAlternative {
        val extractor = patternAnalyzer.createExtractor(alt.pattern, matchedObject, env)
        val body = analyze(alt.value, env)
        return CoreAlternative(extractor.predicate, CoreSequenceExpression.of(extractor.extractor, body))
    }
}
